src/driver/mod.rs

```diff
@@ -238,47 +234,37 @@ impl<E: Engine, S: sequencing::SequencingSource<E>> Driver<E, S> {
             self.future_unsafe_blocks.push(payload);
         }
 
+        let engine_driver = self.engine_driver.read().await;
         self.future_unsafe_blocks.retain(|payload| {
             let unsafe_block_num = payload.block_number.as_u64();
-            let synced_block_num = self.engine_driver.unsafe_head.number;
+            let synced_block_num = engine_driver.unsafe_head.number;
 
             unsafe_block_num > synced_block_num && unsafe_block_num - synced_block_num < 1024
         });
 
         let next_unsafe_payload = self
             .future_unsafe_blocks
             .iter()
-            .find(|p| p.parent_hash == self.engine_driver.unsafe_head.hash);
+            .find(|p| p.parent_hash == engine_driver.unsafe_head.hash);
 
         if let Some(payload) = next_unsafe_payload {
-            _ = self.engine_driver.handle_unsafe_payload(payload).await;
```

Comment:
can we move this to `engine_driver.rs`? like a top-level function called `handle_unsafe_payload`

Reply:
it's a bit awkward to because of where the first read lock is acquired/released, i can try to refactor later by having a function that takes an optional guard